### PR TITLE
feat/#470/publish tailcall on npm

### LINKED ISSUE #470: Publish tailcall on npm
Nothing much to describe here except that we need to publish tailcall on `tailcallhq/tailcall` :) 

**Technical Requirements:**
- CI should be integrated and for every release on github, the package should be published on npm.
- Use `NPM_TOKEN` secret to publish on npm.

**Additional Information:**
~~- Use this as reference: https://github.com/orhun/git-cliff/tree/main/npm~~
- Use this as reference - https://github.com/tusharmath/rust-npm-demo/edit/main/README.md

**IMPORTANT**
PLEASE DON'T ATTEMPT TO CALL A RUST PROGRAM VIA NODE.JS.

### PR DESCRIPTION
**Pull request by ologbonowiwi**
- feat: add basic npm stuff
- feat: add optional dependencies
- build: add npm publishing on ci.yml
- feat: call tailcall binary properly
- build: add npm publish for main/base package

**Summary:**  
Added npm publishing steps on 

**Issue Reference(s):**  
Fixes #470
/claim #470

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh` to address and fix linting issues.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the documentation accordingly (if applicable).
- [ ] I have performed a self-review of my own code.

### CODE REVIEW

**Comment by ologbonowiwi** [comment]
@tusharmath I've used [orhun/git-cliff](https://github.com/orhun/git-cliff) approach as you asked; can you review it?

File: npm/tailcall/package.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "tailcall",
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:2]:
```suggestion
  "name": "@tailcallhq/tailcall",
```

File: npm/package.json.template [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "${node_package}",
```
Comment (ologbonowiwi) [inline npm/package.json.template:2]:
```suggestion
  "name": "@tailcallhq/${node_package}",
```

File: npm/tailcall/package.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/tailcall/tailcall.git"
+  },
+  "homepage": "https://github.com/tailcall/tailcall#readme",
+  "bugs": {
+    "url": "https://github.com/tailcall/tailcall/issues"
+  },
+  "optionalDependencies": {
+    "tailcall-linux-x86_64": "0.1.0",
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:27]:
```suggestion
    "@tailcallhq/tailcall-linux-x86_64": "0.1.0",
```

File: npm/tailcall/package.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "tailcall-linux-aarch64": "0.1.0",
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:28]:
```suggestion
    "@tailcallhq/tailcall-linux-aarch64": "0.1.0",
```

File: npm/tailcall/package.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    "tailcall-linux-i686": "0.1.0",
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:29]:
```suggestion
    "@tailcallhq/tailcall-linux-i686": "0.1.0",
```

File: npm/tailcall/package.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    "tailcall-apple-arm64": "0.1.0",
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:30]:
```suggestion
    "@tailcallhq/tailcall-apple-arm64": "0.1.0",
```

File: npm/tailcall/package.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    "tailcall-apple-x86_64": "0.1.0"
```
Comment (ologbonowiwi) [inline npm/tailcall/package.json:31]:
```suggestion
    "@tailcallhq/tailcall-apple-x86_64": "0.1.0"
```

**Comment by tusharmath** [comment]
@ologbonowiwi We don't intend to run a native program via node.js. NPM can be used to publish any native package.

**Comment by ologbonowiwi** [comment]
But on the original issue, you said
> Use this as reference: https://github.com/orhun/git-cliff/tree/main/npm

And on git-cliff, they call the native program via Node.

What do you expect instead, @tusharmath?

**Comment by ologbonowiwi** [comment]
@tusharmath one possible option would be to use bash instead of Node to interface with the CLI. But in this case, we'll have issues, which will likely bring complications when we use Windows (which does not run shell files).

I understand why you don't have node interfacing tailcall, but it'd be hard not to have any node code since we're trying to publish the package on the node ecosystem here.

The trade-off of having the cli that it'll work on any platform via NPM is to have a dumb layer that runs the binary for you. And I'm sure this Node code runs in less than 1 sec. Having the node layer interface will have great value in terms of interoperability between Unix/Windows/any system that runs Node is a great price to pay for the "easiness" we're having here.

Let me know if we can go on with the current implementations (with any needed tweaks) or if you have any other idea that we can have closer interop (at least we need to think in one option that works also in Windows)

**Comment by tusharmath** [comment]
@ologbonowiwi I think you raised a great point. I went thru it again and realized the approach is absolutely incorrect. I have created a demo here https://github.com/tusharmath/rust-npm-demo

You can install the package like this - `npm i -g rust-npm-demo`

**Pull request closed without merging**